### PR TITLE
[JAX] Add TE examples

### PR DESCRIPTION
**Pull request by jeng1220**
Add following examples:
*  examples/jax/test_single_gpu_mnist.py
*  examples/jax/test_single_gpu_encoder.py
*  examples/jax/test_multigpu_encoder.py
*  examples/jax/test_model_parallel_encoder.py

Align default values to [google/flax](https://github.com/google/flax)
* transformer_engine/jax/module.py
* transformer_engine/jax/transformer.py

### CODE REVIEW

File: tests/jax/test_mnist.py [hunk 1]
```diff
@@ -1,227 +0,0 @@
-# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (nouiz) [inline tests/jax/test_mnist.py:1]: The CI run the JAX example. This is great.
If a user or a dev run TE tests, I think currently the examples won't be run.
Any idea how to make them also run?

File: qa/L0_jax_unittest/test.sh [hunk 2]
```diff
@@ -6,4 +6,6 @@ set -xe
 
 : ${TE_PATH:=/opt/transformerengine}
 pytest -Wignore -v $TE_PATH/tests/jax
+
+pip install -r $TE_PATH/examples/jax/encoder/requirements.txt
```
Comment (nouiz) [inline qa/L0_jax_unittest/test.sh:10]: Right now, it will works as mnist requirements is a subset of encoder requirements.
But to be more future proff, can you also install the mnist requirements file?

File: examples/jax/encoder/test_multigpu_encoder.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+            in_shardings = (state_pspec, inputs_pspec, labels_pspec, None, None)
+            out_shardings = (None, None, None)
+            pjit_apply_model = pjit(apply_model, in_shardings, out_shardings)
+
+            in_shardings = (state_pspec, None)
+            out_shardings = (None, None)
+            pjit_update_model = pjit(update_model, in_shardings, out_shardings, static_argnums=2)
+
+            if args.use_fp8:
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/test_multigpu_encoder.py:283]: every example has this checker

File: tests/jax/test_mnist.py (hunk 1 above)
Comment (jeng1220) [reply tests/jax/test_mnist.py:1]: It should work. User can follow `qa/L0_jax_unittest/test.sh` to run both unittests and examples

File: examples/jax/encoder/test_multigpu_encoder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nouiz) [reply examples/jax/encoder/test_multigpu_encoder.py:283]: I seem them today. Not sure why I didn't saw them yesterday.

File: tests/jax/test_mnist.py (hunk 1 above)
Comment (nouiz) [reply tests/jax/test_mnist.py:1]: The normal workflow to run tests is to run:
`pytest directory`

So I'm sure some users will skip the documentation on how to run the tests.
If by chance you have an idea how to fix that, that is great. Otherwise, it can stay like this. Check if there is a section on how to run tests and update it.

File: examples/jax/encoder/README.md [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Basic Transformer Encoder Example with Optional FP8
+
+```bash
+python test_single_gpu_encoder.py               # Encoder from TransformerEngine
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/README.md]: The README.md were updated.

File: examples/jax/encoder/test_multigpu_encoder.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+    state_pspec = jax.tree_map(replace_params, state, is_leaf=lambda x: isinstance(x, FrozenDict))
+    return state_pspec
+
+
+def train_and_evaluate(args):
+    """Execute model training and evaluation loop."""
+    print(args)
+    assert num_gpu_is_enough(), "Need 8 GPUs to run"
+    assert args.batch_size % NUM_GPU == 0, "Batch size needs to be multiple of 8"
```
Comment (nouiz) [inline examples/jax/encoder/test_multigpu_encoder.py:240]: Replace `"Batch size needs to be multiple of 8"`
with `"Batch size needs to be multiple of " + NUM_GPU`
or `"Batch size needs to be multiple of the number of GPUs"`

File: examples/jax/encoder/test_model_parallel_encoder.py [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+                            kernel_axes=(NAMED_TP_AXIS, NAMED_BROADCAST_AXIS),
+                            bias_axes=(NAMED_BROADCAST_AXIS,),
+                            sharding_type=te.ShardingType.DP_TP_ROW,
+                            dtype=jnp.bfloat16)(x)
+
+        x = nn.Dense(features=2, dtype=jnp.bfloat16)(x)
+        return x
+
+
+def train_step(state, inputs, labels, var_collect, rngs, use_fp8):
```
Comment (nouiz) [inline examples/jax/encoder/test_model_parallel_encoder.py:94]: Miss `@partial(jax.jit, static_argnums=5)` from the singlegpu codebase.

File: examples/jax/encoder/test_multigpu_encoder.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+
+        x = te.DenseGeneral(features=256, sharding_type=te.ShardingType.DP, dtype=jnp.bfloat16)(x)
+
+        x = te.DenseGeneral(features=256, sharding_type=te.ShardingType.DP, dtype=jnp.bfloat16)(x)
+
+        x = nn.Dense(features=2, dtype=jnp.bfloat16)(x)
+        return x
+
+
+def train_step(state, inputs, labels, var_collect, rngs, use_fp8):
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/test_multigpu_encoder.py:80]: This example uses `pjit` to set up complex `PartitionSpec` instead of `jax.jit`
```python
            in_shardings = (state_pspec, inputs_pspec, labels_pspec, None, None)
            out_shardings = (state_pspec, None, None, None)
            pjit_train_step = pjit(train_step, in_shardings, out_shardings, static_argnums=(5,))
```

File: examples/jax/encoder/test_multigpu_encoder.py [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+
+class Net(nn.Module):
+    """NLP Encoder"""
+    num_embed: int
+
+    @nn.compact
+    def __call__(self, x, disable_dropout=False):
+        x = nn.Embed(num_embeddings=self.num_embed, features=256, dtype=jnp.bfloat16)(x)
+
+        te_Encoder = partial(te.TransformerLayer,
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/test_multigpu_encoder.py:58]: Yes, it does.
I forgot to describe it

File: examples/jax/encoder/README.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+python test_multigpu_encoder.py --use-fp8
+```
+
+## Multiple GPU with Model Parallelism ##
+
+1. The model parallelism as known as tensor parallelism (TP) divides a model for multiple devices, and each device has part of model parameters. This example inherits previous DP example, but divides a model to two devices.
+
+2. To set up device mesh for TP, adding a new named axis called `model`, which is used for sharding parameters of the model across devices.
+
+3. On the model side, the first `te.DenseGeneral` is divided by columns and second one is divided by rows for TP. Because `te.DenseGeneral` doesn't have the default named axis, the names must be set manually by passing `kernel_axes` and `bias_axes` arguments. Then, the rest of the workflow is similar to the previous example.
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/README.md:53]: Will update README.md later

File: examples/jax/mnist/test_single_gpu_mnist.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+    @unittest.skipIf(not gpu_has_fp8(), reason='GPU capability is not enough to run FP8')
+    def test_te_fp8(cls):
+        """Test Transformer Engine with FP8"""
+        cls.args.use_fp8 = True
+        actual = train_and_evaluate(cls.args)
+        cls.verify(actual)
+
+
+if __name__ == "__main__":
+    train_and_evaluate(mnist_parser(None))
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/mnist/test_single_gpu_mnist.py:309]: It is correct.
```python
mnist_parser(["--epochs", "3"]) # manually set up arguments in the code,
                                # and avoid conlict with pytest's arguments 
```
vs.
```python
mnist_parser(None) # get arguments from command line
```

File: tests/jax/test_mnist.py (hunk 1 above)
Comment (jeng1220) [reply tests/jax/test_mnist.py:1]: > TE already document this?

No, it doesn't

File: qa/L0_jax_unittest/test.sh (hunk 2 above)
Comment (jeng1220) [reply qa/L0_jax_unittest/test.sh:11]: added

File: tests/jax/test_mnist.py (hunk 1 above)
Comment (jeng1220) [reply tests/jax/test_mnist.py:1]: It needs to open another thread for discussion.

File: examples/jax/mnist/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+5. Evaluating process: The evaluating process is the same as the training process. Need to ensure FP8 metadata is inside var_collect and fill it into loss function.
+
+6. Additional options: The `te.fp8_autocast` context manager has additional options
+   * FP8 Recipe: control FP8 training behavior. See the [FP8 tutorial](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html) for a detailed explanation of FP8 recipes and the supported options. **Noted** that FP8 metadata is now the responsibility of the user to update because of JAX limitation (i.e., manually calling `te.update_fp8_metas`). The JAX version of Transformer Engine cannot update FP8 metadata on its own.
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/mnist/README.md:16]: > Does the PyTorch can do it?

PyToch can do it.

`because of JAX limitation` was removed.

For hiding FP8 metadata update into train_state, it is a good point, but we are running out of the time. It can be another thread.

File: examples/jax/mnist/requirements.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+tensorflow-datasets
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/mnist/requirements.txt:2]: added optax.

For Flax, the Transformer Engine is built on top of Flax, so if user doesn't have Flax, he/she cannot even build Transformer Engine.

File: examples/jax/mnist/README.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+   * Sharding Resource: tell Transformer Engine how to make data parallelism and tensor parallelism. We will introduce it more in Encoder examples.
+
+## Run ##
+
+1. Use Flax to train MNIST as usual
+```bash
+python test_single_gpu_mnist.py
+```
+
+2. Use `te.DenseGeneral` provided by Transformer Engine to train MNISt with BF16
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/mnist/README.md:26]: @nouiz ,
What is it? I do not understand what you mean

File: examples/jax/encoder/test_multigpu_encoder.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/test_multigpu_encoder.py:59]: updated README.md

File: examples/jax/encoder/test_multigpu_encoder.py (hunk 5 above)
Comment (jeng1220) [reply examples/jax/encoder/test_multigpu_encoder.py]: done

File: examples/jax/encoder/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/README.md]: done

File: examples/jax/mnist/test_single_gpu_mnist.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+        "--dry-run",
+        action="store_true",
+        default=False,
+        help="quickly check a single pass",
+    )
+    parser.add_argument("--seed", type=int, default=1, metavar="S", help="random seed (default: 1)")
+    parser.add_argument("--use-fp8",
+                        action="store_true",
+                        default=False,
+                        help="Use FP8 for inference and training without recalibration")
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/mnist/test_single_gpu_mnist.py]: done

File: examples/jax/encoder/README.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+### Run ###
+
+```bash
+python test_single_gpu_encoder.py
+python test_single_gpu_encoder.py --use-fp8
+```
+
+## Multiple GPU with Data Parallelism ##
+
+1. The data parallelism (DP) divides a mini-batch for multiple devices, and each device has complete model parameters. In this example, the first dimension of input tensor is `batch_size` which is 64 by default, and uses 8 GPUs to train the model, so each device takes 8 sentences at once. The "dividing" is also called "sharding" in the JAX documents.
```
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/encoder/README.md]: done

File: examples/jax/mnist/README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (jeng1220) [reply examples/jax/mnist/README.md]: oh, I got it.
README.md was updated

**Comment by jeng1220** [comment]
@timmoon10 and @nouiz ,
I think all conversation are resolved.
And I saw internal pipeline 7708362 is good.

**Comment by nouiz** [comment]
> @timmoon10 and @nouiz , I think all conversation are resolved. And I see internal pipeline 7708362 is good.

For me it is good except for one thing. GH doesn't allow me to make the comment at the right place, so trying here.

For the requirement, we still need to add flax as we don't always build and install in the same env.
It happens frequently that people use what is on pypi or build and one env and reuse in another one (We do this frequently in Dockerfile).

So we also need to add flax, even if in some cases it won't be useful.

**Comment by jeng1220** [comment]
> For me it is good except for one thing. GH doesn't allow me to make the comment at the right place, so trying here.
> 
> For the requirement, we still need to add flax as we don't always build and install in the same env. It happens frequently that people use what is on pypi or build and one env and reuse in another one (We do this frequently in Dockerfile).
> 
> So we also need to add flax, even if in some cases it won't be useful.

Done

**Comment by nouiz** [comment]
What is the result of the CI?
What is blocking this PR from being merged?

**Review by timmoon10 — APPROVED** [review]
LGTM

The CI is waiting on some A100 tests, but everything else has passed.

**Comment by jeng1220** [comment]
@timmoon10 and @nouiz ,
Thanks for help
